src/d/actor/d_a_obj_barrel.cpp

```diff
@@ -4,226 +4,758 @@
 void daObjBarrel::Act_c::init_mtx() {
-    /* Nonmatching */
+    mpModel->setBaseScale(scale);
+    set_mtx();
 }
 
 /* 000014CC-00001710       .text set_walk_rot__Q211daObjBarrel5Act_cFv */
 void daObjBarrel::Act_c::set_walk_rot() {
-    /* Nonmatching */
+    cXyz delta = current.pos - old.pos;
+    float mag = delta.absXZ();
+
+    bool negAngle = false;
+    current.angle.y = cM_atan2s(delta.x, delta.z);
+    short targetAngle = current.angle.y;
+    if (targetAngle < 0) {
+        targetAngle += 0x8000;
+        negAngle = true;
+    }
+    if (mag > l_min_move_dir || mMode == MODE_WAIT && mag > l_min_move_dir / 2) {
+        cLib_chaseAngleS(&shape_angle.y, targetAngle, 0x600);
+    }
+    float fVar2 = mag / ((cM_scos(shape_angle.z) * 5.0f + l_s_radius) * 6.28f) * 65535.0f;
+    if (!negAngle) {
+        m612 -= (short)(fVar2 * 3.0f);
+        m630 -= (short)fVar2;
+    } else {
+        m612 += (short)(fVar2 * 3.0f);
+        m630 += (short)fVar2;
+    }
+    float vibFactor = mag / l_max_move;
+    if (vibFactor > 1.0f) {
+        vibFactor = 1.0f;
+    }
+    m610 = vibFactor * l_max_vib_angl;
 }
 
 /* 00001710-00001824       .text eff_break__Q211daObjBarrel5Act_cFv */
 void daObjBarrel::Act_c::eff_break() {
-    /* Nonmatching */
+    cXyz pos(current.pos.x, current.pos.y + 50.0f, current.pos.z);
+    JPABaseEmitter* emitter = dComIfGp_particle_set(dPa_name::ID_COMMON_03E5, &pos, NULL, NULL, 0xFF, NULL, -1, &tevStr.mColorK0, &tevStr.mColorK0);
+    if (emitter) {
+        static const JGeometry::TVec3<f32> em_scl(1.0f, 0.8f, 1.0f);
+        emitter->setEmitterScale(em_scl);
+    }
+    daObjEff::Act_c::make_barrel_smoke(&pos);
 }
 
 /* 00001824-000018C8       .text damaged__Q211daObjBarrel5Act_cFb */
-void daObjBarrel::Act_c::damaged(bool) {
-    /* Nonmatching */
+void daObjBarrel::Act_c::damaged(bool animate) {
+    fopAcM_cancelCarryNow(this);
+    if (animate) {
+        eff_break();
+        fopAcM_seStart(this, JA_SE_OBJ_COL_SWC_WODCRL, 0);
+        set_senv(attr().mSeBreakP1, attr().mSeBreakP2);
+    }
 }
 
 /* 000018C8-00001900       .text cull_set_draw__Q211daObjBarrel5Act_cFv */
 void daObjBarrel::Act_c::cull_set_draw() {
-    /* Nonmatching */
+    fopAcM_setCullSizeSphere(this, 0.0f, 50.0f, 0.0f, 75.0f);
 }
 
 /* 00001900-00001938       .text cull_set_move__Q211daObjBarrel5Act_cFv */
 void daObjBarrel::Act_c::cull_set_move() {
-    /* Nonmatching */
+    fopAcM_setCullSizeSphere(this, 0.0f, 50.0f, 0.0f, 300.0f);
 }
 
 /* 00001938-00001940       .text get_se_map_hit__Q211daObjBarrel5Act_cCFv */
-void daObjBarrel::Act_c::get_se_map_hit() const {
-    /* Nonmatching */
+u32 daObjBarrel::Act_c::get_se_map_hit() const {
+    return 0x0B;
 }
 
 /* 00001940-00001994       .text set_senv__Q211daObjBarrel5Act_cCFii */
-void daObjBarrel::Act_c::set_senv(int, int) const {
-    /* Nonmatching */
+void daObjBarrel::Act_c::set_senv(int p1, int p2) const {
+    // TODO:  create cXyz?
```

Comment:
Old comment?

Reply:
I noticed a call to `__ct__4cXyzFRC4cXyz` in the debug map for this function before the `fopAcM_GetID`, but not sure what it's for. I'll remove the comment though

Reply:
It's because dKy_Sound_set takes a position passed by value and not reference, so when you pass `current.pos` the compiler automatically creates a new cXyz on the stack using the copy constructor `cXyz::cXyz(const cXyz&)`, and then passes a pointer to the temp on the stack to the function. This gets inlined but if you want to see what it would have looked like for reference, throw `#pragma dont_inline on` above this function and you should see the call.